src/app/loadout/mod-utils.ts

```diff
@@ -325,62 +327,85 @@ function createOrderedAssignmentResults(
+    // If it wasn't found already plugged, find the first socket it can fit into
+    if (destinationSocketIndex === -1) {
+      destinationSocketIndex = existingModSockets.findIndex(
         (socket) => socket.plugged?.plugDef.plug.plugCategoryHash === mod.plug.plugCategoryHash
       );
     }
 
-    // This should always be the case, if its not then something is seriously wrong
-    if (sortedSocketIndex > 0) {
-      const pluggedCost =
-        modSockets[sortedSocketIndex].plugged?.plugDef.plug.energyCost?.energyCost || 0;
-      const modCost = mod.plug.energyCost?.energyCost || 0;
-      const energyChange = modCost - pluggedCost;
-      results.push({ socketIndex: modSockets[sortedSocketIndex].socketIndex, mod, energyChange });
-      modSockets.splice(sortedSocketIndex, 1);
+    // If a destination socket couldn't be found for this plug, something is seriously wrong
+    if (destinationSocketIndex === -1) {
+      throw new Error(
+        `We couldn't find anywhere to plug the mod ${mod.displayProperties.name} (${mod.hash})`
+      );
     }
+
+    const existingModCost =
+      existingModSockets[destinationSocketIndex].plugged?.plugDef.plug.energyCost?.energyCost || 0;
+    const plannedModCost = mod.plug.energyCost?.energyCost || 0;
+    const energyChange = plannedModCost - existingModCost;
+
+    pluggingActions.push({
+      socketIndex: existingModSockets[destinationSocketIndex].socketIndex,
+      mod,
+      energyChange,
+    });
+
+    // remove this existing socket from consideration
+    existingModSockets.splice(destinationSocketIndex, 1);
   }
 
-  // Add in the default plug of each socket to anything that didn't get a mod
-  for (const leftoverSocket of modSockets) {
-    const mod = defs?.InventoryItem.get(leftoverSocket.socketDefinition.singleInitialItemHash);
+  // For each remaining socket that won't have mods assigned,
+  // return it to its default (usually "Empty Mod Socket")
+  for (const leftoverSocket of existingModSockets) {
+    const defaultMod = defs?.InventoryItem.get(
+      leftoverSocket.socketDefinition.singleInitialItemHash
+    );
     const currentModEnergy = leftoverSocket.plugged?.plugDef.plug.energyCost?.energyCost || 0;
-    results.push({
+    pluggingActions.push({
       socketIndex: leftoverSocket.socketIndex,
-      mod: mod as PluggableInventoryItemDefinition,
+      mod: defaultMod as PluggableInventoryItemDefinition,
       energyChange: -currentModEnergy,
     });
   }
 
-  // Sort so we take away energy before we add it if possible
-  return results.sort(compareBy((res) => res.energyChange));
+  // Sort so we free up energy before we consume it, if possible
+  return pluggingActions.sort(compareBy((res) => res.energyChange));
 }
 
+/**
+ * input: a dictionary keyed by instanceId, of `{ socketIndex, mod|undefined}[]`
+ *
+ * output: a dictionary keyed by instanceId, of `mod[]`
+ *
+ * used to make a series of mod assignments slot-agnostic?
```

Comment:
Yeah this was basically just the easiest way to keep the existing UI functionality. Ideally we will remove this function and just use the assignments as is, but that was a bunch more work I didn't want to do at 1am